src/core/api/index.ts

```diff
@@ -711,16 +744,28 @@ class Player extends EventEmitter<any> {
      */
     const stateChanges$ = loaded.mapTo(PLAYER_STATES.LOADED)
       .concat(
-        Observable.combineLatest(this._priv_playing$, stalled$)
-          .takeUntil(this._priv_stopCurrentContent$)
-          .map(([isPlaying, stalledStatus]) => {
-            if (stalledStatus) {
-              return (stalledStatus.state === "seeking") ?
-                PLAYER_STATES.SEEKING : PLAYER_STATES.BUFFERING;
-            }
-            return isPlaying ? PLAYER_STATES.PLAYING : PLAYER_STATES.PAUSED;
-          })
-
+        Observable.merge(
+          Observable.combineLatest(this._priv_playing$, stalled$)
+            .takeUntil(this._priv_stopCurrentContent$)
+             .map(([isPlaying, stalledStatus]) => {
+               if (stalledStatus) {
+                 switch(stalledStatus.state){
+                   case "seeking":
+                     return PLAYER_STATES.SEEKING;
+                   default:
+                    return PLAYER_STATES.BUFFERING;
+                }
+              }
+              return isPlaying ? PLAYER_STATES.PLAYING : PLAYER_STATES.PAUSED;
+            })
+            .distinctUntilChanged(),
+          this._priv_seeked$
+            .map((isPlaying) => {
+              return isPlaying ? PLAYER_STATES.PLAYING : PLAYER_STATES.PAUSED;
+            }),
+          this._priv_ended$
+            .mapTo(PLAYER_STATES.ENDED)
+        )
```

Comment:
Too unreadable :/

Why not using the already defined combineLatest here for ended?

+ seeked looks a little complicated there